File: wireguard_plugin/android/src/main/kotlin/net/defguard/wireguard_plugin/TunnelStartData.kt
package net.defguard.wireguard_plugin

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class TunnelTraffic {
    @SerialName("all")
    ALL,

    @SerialName("predefined")
    PREDEFINED,
}

@Serializable
data class TunnelStartData(
    // config
    val publicKey: String,
    val privateKey: String,
    val address: String,
    val dns: String? = null,
    val endpoint: String,
    val allowedIps: String,
    val keepalive: Int,
    val presharedKey: String?,
    val traffic: TunnelTraffic,
    // context
    val locationName: String,
    val locationId: Int,
    val instanceId: Int
)

File: client/ios/VPNExtension/TunnelStartData.swift
//
//  TunnelStartData.swift
//  Runner
//
//  Created by Aleksander on 08/07/2025.
//

import Foundation

public enum TunnelTraffic: String, Codable {
    case All = "all"
    case Predefined = "predefined"
}

public struct TunnelStartData: Codable {
    public var publicKey: String
    public var privateKey: String
    public var address: String
    public var dns: String?
    public var endpoint: String
    public var allowedIps: String
    public var keepalive: Int
    public var presharedKey: String?
    public var traffic: TunnelTraffic
    public var locationName: String
    public var locationId: Int
    public var instanceId: Int
    
    public func endpointPort() -> Int? {
        guard let url = URL(string: endpoint) else { return nil }
        return url.port
    }
    
    public func endpointHost() -> String? {
        guard let url = URL(string: endpoint) else { return nil }
        return url.host
    }
}


struct ActiveTunnelData: Codable {
    var locationId: Int
    var instanceId: Int
    var traffic: TunnelTraffic

    init(fromConfig: TunnelStartData) {
        self.locationId = fromConfig.locationId
        self.instanceId = fromConfig.instanceId
        self.traffic = fromConfig.traffic
    }
}
